Modified altdiffusion pipline to support altdiffusion-m18

superhero-7 opened the pull request:
Hi, friends at huggingface. I am from BAAI, our altdiffusion have been integrated in diffusers before. Recently, we release a new version of altdiffusion, named altdiffusion-m18, which supports 18 lanugages. And there has some little different from previous altdiffusion, so we modified the pipeline file. And we hope you can provide some advice of code compatibility, Thaks!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2993). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/modeling_roberta_series.py:
```diff
@@ -65,6 +67,8 @@ def __init__(self, config):
         super().__init__(config)
         self.roberta = XLMRobertaModel(config)
         self.transformation = nn.Linear(config.hidden_size, config.project_dim)
+        self.transformation_pre = nn.Linear(config.hidden_size, config.project_dim)
```
We should only add those under the condition that it's a m18 checkpoint otherwise this breaks for previous checkpoints. Could we add a `config.has_pre_transformation` flag?

superhero-7 replied:
> We should only add those under the condition that it's a m18 checkpoint otherwise this breaks for previous checkpoints. Could we add a `config.has_pre_transformation` flag?

Exactly! I also think so, but I don't know where to add this flag properly, so I am looking for your help.

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/modeling_roberta_series.py:
```diff
@@ -95,14 +99,21 @@ def forward(
             output_attentions=output_attentions,
-            output_hidden_states=output_hidden_states,
+            # output_hidden_states=output_hidden_states,
+            output_hidden_states=True,
```
Only necessary if we're in m18 mode - we still need to support the other use case

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:
```diff
@@ -313,10 +315,6 @@ def _encode_prompt(
             batch_size = prompt_embeds.shape[0]
 
         if prompt_embeds is None:
-            # textual inversion: procecss multi-vector tokens if necessary
```
Why delete this?

superhero-7 replied:
> Why delete this?
Sorry for that. I think I deleted it by mistake when I used black to format the code.

patrickvonplaten commented:
Hey @superhero-7,

The PR looks very cool! Could you maybe just make sure that the old checkpoints are not broken - we could achieve this by adding a flag to the config :-)

superhero-7 commented:
Hi, according to your suggestion, I re-edited the pr. Compared with the code in the current diffuser library, only minor modifications have been made in the file modeling_roberta_series.py, keeping pipeline_alt_diffusion.py unchanged. But the pipeline_alt_diffusion.py file I haven't modified but can't pass the code test, I find it very strange！Can you take a look and give some advice, thanks！

superhero-7 commented:
Hi, friends at huggingface. I think just add a flag in textencoder config, and modified  modeling_roberta_series.py will be ok. The unsuccessful checks right now I believe is due to the checkpoint for test don't add the has_pre_transformation flag. I have already added the flag in m9 and m18 modelhub, but I don't if the checkpoint for test is the same as modelhub? I am looking forward your advice, thanks a lot!

superhero-7 commented:
@patrickvonplaten Sorry to bother, any suggestions about this pr?

patrickvonplaten commented:
Hey @superhero-7 three tests are currently failing due to this PR: 
```
FAILED tests/pipelines/altdiffusion/test_alt_diffusion.py::AltDiffusionPipelineFastTests::test_alt_diffusion_ddim - AttributeError: 'RobertaSeriesConfig' object has no attribute 'has_pre_transformation'
FAILED tests/pipelines/altdiffusion/test_alt_diffusion.py::AltDiffusionPipelineFastTests::test_alt_diffusion_pndm - AttributeError: 'RobertaSeriesConfig' object has no attribute 'has_pre_transformation'
FAILED tests/pipelines/altdiffusion/test_alt_diffusion_img2img.py::AltDiffusionImg2ImgPipelineFastTests::test_stable_diffusion_img2img_default_case - AttributeError: 'RobertaSeriesConfig' object has no attribute 'has_pre_transformation'
```

Can you try to make sure to not introduce breaking changes in the PR while extending it to a new model? :-) Leaving more comments directly in the code

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/modeling_roberta_series.py:
```diff
@@ -65,6 +65,10 @@ def __init__(self, config):
         super().__init__(config)
         self.roberta = XLMRobertaModel(config)
         self.transformation = nn.Linear(config.hidden_size, config.project_dim)
+        self.has_pre_transformation = config.has_pre_transformation
```
```suggestion
        self.has_pre_transformation = getattr(config, "has_pre_transformation", False)
```

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/modeling_roberta_series.py:
```diff
@@ -85,25 +89,51 @@ def forward(
+                encoder_hidden_states=encoder_hidden_states,
+                encoder_attention_mask=encoder_attention_mask,
+                output_attentions=output_attentions,
+                output_hidden_states=True,
```
```suggestion
                output_hidden_states=True if self.has_pre_transformation else output_hidden_states,
```
and then we can probably remove the if-else statement completely :-)

patrickvonplaten commented on src/diffusers/pipelines/alt_diffusion/modeling_roberta_series.py:
```diff
@@ -85,25 +89,51 @@ def forward(
+                return_dict=return_dict,
+            )
+        else:
+            outputs = self.base_model(
```
This can be simplified if we do: https://github.com/huggingface/diffusers/pull/2993/files#r1164511366 :-)

patrickvonplaten commented:
Think it's an easy fix to get the tests passing - just these two comments here: 
- https://github.com/huggingface/diffusers/pull/2993#discussion_r1164511366
- https://github.com/huggingface/diffusers/pull/2993#discussion_r1164511778

superhero-7 commented:
@patrickvonplaten Thanks a lot!!! I modified the code according your suggestion, and pass all of the check! Could you please have a look again?~Thanks !!!
